deeplearning4j-cuda/src/test/java/org/deeplearning4j/gradientcheck/CNNGradientCheckTest.java

```diff
@@ -926,25 +926,89 @@ public void testDeconvolution2D() {
         }
     }
 
+    @Test
+    public void testDepthwiseConv2D() {
+        int nIn = 3;
+        int depthMultiplier = 2;
+        int nOut = nIn * depthMultiplier;
+
+        int width = 8;
+        int height = 8;
+
+        Nd4j.getRandom().setSeed(12345);
+
+        int[] ks = new int[]{1,3,1,3,1,3,1,3};
+        int[] ss = new int[]{1,1,2,2,1,1,2,2};
+        ConvolutionMode[] cms = new ConvolutionMode[]{
+                Truncate, Truncate, Truncate, Truncate, Truncate, Truncate, Truncate, Truncate};
+        int[] mb = new int[]{1,1,3,3,3,1,3,3};
+
+        for( int t=0; t<ks.length; t++ ){
+
+            int k = ks[t];
+            int s = ss[t];
+            ConvolutionMode cm = cms[t];
+            int minibatchSize = mb[t];
+
+
+            INDArray input = Nd4j.rand(minibatchSize, width * height * nIn);
+            INDArray labels = Nd4j.zeros(minibatchSize, nOut);
+            for (int i = 0; i < minibatchSize; i++) {
+                labels.putScalar(new int[]{i, i % nOut}, 1.0);
+            }
+
+            NeuralNetConfiguration.ListBuilder b = new NeuralNetConfiguration.Builder().seed(12345)
+                    .updater(new NoOp())
+                    .activation(Activation.TANH)
+                    .convolutionMode(cm)
+                    .list()
+                    .layer(new DepthwiseConvolution2D.Builder().name("depth-wise conv 2D layer")
+                            .cudnnAllowFallback(false)
+                            .kernelSize(k, k)
+                            .stride(s, s)
+                            .depthMultiplier(depthMultiplier)
+                            .nIn(nIn).build()); // nOut = nIn * depthMultiplier
```

Comment:
We need a layer (with parameters) below the depthwise conv - say a standard conv layer, though it doesn't really mattech what - for the test to be sufficiently thorough.
Without a layer below it, the epsilons (activation gradients) out of depthwise conv could be wrong and yet the test would still pass.